src/sentry/sentry_metrics/multiprocess.py

```diff
@@ -0,0 +1,334 @@
+@functools.lru_cache(maxsize=10)
+def get_indexer():
+    from sentry.sentry_metrics import indexer
+
+    return indexer
+
+
+def process_messages(outer_message: Message[TPayload]) -> List[Message[TPayload]]:
+    """
+    We have an outer_message Message() whose payload is a batch of Message() objects.
+
+        Message(
+            partition=...,
+            offset=...
+            timestamp=...
+            payload=[Message(...), Message(...), etc]
+        )
+
+    The inner messages payloads are KafkaPayload's that have:
+        * key
+        * headers
+        * value
+
+    The value of the message is what we need to parse and then translate
+    using the indexer.
+    """
+    indexer = get_indexer()
+
+    strings = set()
+    parsed_payloads_by_offset = {
+        msg.offset: json.loads(msg.payload.value, use_rapid_json=True)
+        for msg in outer_message.payload
+    }
+
+    for message in parsed_payloads_by_offset.values():
+        metric_name = message["name"]
+        tags = message.get("tags", {})
+
+        parsed_strings = {
+            metric_name,
+            *tags.keys(),
+            *tags.values(),
+        }
+        strings.update(parsed_strings)
+
+    mapping = indexer.bulk_record(list(strings))  # type: ignore
+
+    new_messages: Sequence[Message[KafkaPayload]] = []
+
+    for message in outer_message.payload:
+        parsed_payload_value = parsed_payloads_by_offset[message.offset]
+        new_payload_value = parsed_payload_value
+
+        metric_name = parsed_payload_value["name"]
+        tags = parsed_payload_value.get("tags", {})
+
+        new_tags = {mapping[k]: mapping[v] for k, v in tags.items()}
+
+        new_payload_value["tags"] = new_tags
+        new_payload_value["metric_id"] = mapping[metric_name]
+        new_payload_value["retention_days"] = 90
+
+        new_payload = KafkaPayload(
+            key=message.payload.key,
+            value=json.dumps(new_payload_value).encode(),
+            headers=message.payload.headers,
+        )
+        new_message = Message(
+            partition=message.partition,
+            offset=message.offset,
+            payload=new_payload,
+            timestamp=message.timestamp,
+        )
+        new_messages.append(new_message)
+
+    # TODO(meredith): Need to add kickin off the celery task in here, eventually
+    # we will use kafka to forward messages to the product data model
+    # messages = [
+    #     {"tags": m.payload["tags"], "name": m.payload["name"], "org_id": m.payload["org_id"]}
+    #     for m in self.__messages
+    # ]
+    # process_indexed_metrics.apply_async(kwargs={"messages": messages})
+
+    return new_messages
+
+
+class MetricsConsumerStrategyFactory(ProcessingStrategyFactory):
+    def __init__(
+        self,
+        process_messages: Callable[[Message[TPayload]], TProcessed],
+        next_step: Callable[
+            [Callable[[Mapping[Partition, Position]], None]], ProcessingStrategy[TProcessed]
+        ],
```

Comment:
Why do we need to provide the function to process message and to produce messages to the factory? They will always be the same for the `MetricsConsumerStrategyFactory`